### PR TITLE
[Part 2] Data integrity support in delete queryset, judgement and searchconfig

### PR DESCRIPTION
**Pull request by vibrantvarun**
### Description
This is part 2 of Data integrity support. This PR adds data integrity during deletion of querysets, judgement and search configuration that if they are binded to an experiment then it will throw an exception.

### Issues Resolved
[238](https://github.com/opensearch-project/search-relevance/issues/238)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/searchrelevance/dao/ExperimentDao.java [hunk 1]
```diff
@@ -118,6 +118,27 @@ public SearchResponse getExperiment(String experimentId, ActionListener<SearchRe
...
+            listener.onFailure(new SearchRelevanceException("fieldId cannot be null or empty", RestStatus.BAD_REQUEST));
+            return;
+        }
+
+        if (fieldName == null || fieldName.isEmpty()) {
+            listener.onFailure(new SearchRelevanceException("fieldName cannot be null or empty", RestStatus.BAD_REQUEST));
+            return;
+        }
+
+        SearchSourceBuilder sourceBuilder = new SearchSourceBuilder().query(QueryBuilders.matchQuery(fieldName, fieldId)).size(size);
```
Comment (fen-qin) [inline src/main/java/org/opensearch/searchrelevance/dao/ExperimentDao.java]: shall we use `SearchSourceBuilder sourceBuilder = new SearchSourceBuilder().query(QueryBuilders.termQuery(fieldName, fieldId)).size(size);` instead ?

Comment (vibrantvarun) [reply src/main/java/org/opensearch/searchrelevance/dao/ExperimentDao.java]: I used term query earlier but it was not working as expected. Instead, it was returning 0 results. But match query is working perfectly.

File: src/main/java/org/opensearch/searchrelevance/transport/judgment/DeleteJudgmentTransportAction.java [hunk 2]
```diff
@@ -44,7 +59,31 @@ protected void doExecute(Task task, OpenSearchDocRequest request, ActionListener
                 listener.onFailure(new SearchRelevanceException("judgmentId cannot be null or empty", RestStatus.BAD_REQUEST));
                 return;
             }
-            judgmentDao.deleteJudgment(judgmentId, listener);
+            experimentDao.getExperimentByFieldId(judgmentId, JUDGMENT_LIST, 3, new ActionListener<>() {
```
[the remark this answers is not in the document]
Comment (vibrantvarun) [reply src/main/java/org/opensearch/searchrelevance/transport/judgment/DeleteJudgmentTransportAction.java]: If UserB gets document not found then then the following condition if(expriments.getHits.getTotalHits will be 0. And it will try to delete judgement. 

If document is not found then there will be no error. Instead the hits will be 0. This is handled in the following condition.

File: src/main/java/org/opensearch/searchrelevance/dao/ExperimentDao.java (hunk 1 above)
Comment (vibrantvarun) [reply src/main/java/org/opensearch/searchrelevance/dao/ExperimentDao.java]: The reason it is happening is because we have taken every field in experiment of `keyword` type.

Comment (vibrantvarun) [reply src/main/java/org/opensearch/searchrelevance/dao/ExperimentDao.java]: Yes I know why it happens. Because when we query the experiment index directly by term query, it does not analyze  behind the scenes. For example ab-bc-cd id becomes 
ab , bc, cd  while indexing. So when we try to search with a queryId then ab-bc-cd then it is not able to find. On the other hand, the match query uses analyzer for on both ends the query part as well as the results are already tokenized by OpenSearch during indexing time. That is reason the match query works and terms doesn't.

Note this behaviour is happening because it has text field.
TLDR
OpenSearch uses analyzers by default.

Term query does not uses analyzers unless explicitly set to otherwise. So it tries to look for exact term in the inverted index. Because the mapping is of text type so opensearch while indexing has broken that id to smaller terms using standard analyzer. So that queryId will not be found by term query.

Match query analysis the search query so it is able to find the document.

**Comment by vibrantvarun** [comment]
@martin-gaievski @epugh could you review the PR

**Review by epugh — APPROVED** [review]
Thanks for adding the extra cascade.   Lgtm.

File: src/main/java/org/opensearch/searchrelevance/dao/ExperimentDao.java (hunk 1 above)
Comment (martin-gaievski) [reply src/main/java/org/opensearch/searchrelevance/dao/ExperimentDao.java]: It can be the problem on the opposite side - we may found and delete experiments that actually shouldn't be deleted.
with match instead of term query, there is a small chance of hitting edge case. Example of such scenario: our experiment has id is `prod-search-v1`, and to test few tweaks we create experiment with id `prod-search-v1-optimized`. Match query will return both `prod-search-v1` and `prod-search-v1-optimized`.
